File: Data Structures & Algorithms/subsets-ii/submission-2.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        
        def generateSubset(i, curr):
            if i >= len(nums):
                res.append(curr[:])
                return

            curr.append(nums[i])
            generateSubset(i + 1, curr)
            curr.pop()
            while i < len(nums) - 1 and nums[i] == nums[i + 1]:
                i += 1
            generateSubset(i + 1, curr)

        generateSubset(0, [])
        return res

File: Data Structures & Algorithms/permutations/submission-2.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     * arr.toSpliced(2, 0, "Lene");
     */
    permute(nums) {
        const result = []
        const used = new Array(nums.length).fill(false)

        function dfs(curr) {
            if (curr.length == nums.length) {
                result.push([...curr])
                return
            }

            for (let i = 0; i < nums.length; i++) {
                if (used[i]) continue

                curr.push(nums[i])
                used[i] = true

                dfs(curr)

                curr.pop()
                used[i] = false
            }
        }
        dfs([])
        return result
    }
}
